### PR TITLE
polygon plugin

### PR DESCRIPTION
**Pull request by rocketbitz**
A plugin for polygon's data API (https://polygon.io)

### CODE REVIEW

File: contrib/polygon/api/api.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
+package api
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	nats "github.com/nats-io/go-nats"
```
Comment (umitanuki) [inline contrib/polygon/api/api.go]: Add dep?

Comment (rocketbitz) [reply contrib/polygon/api/api.go]: .lock file is updated.

Comment (umitanuki) [reply contrib/polygon/api/api.go]: Oh i see

File: contrib/polygon/api/api.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+func Stream(handler func(m *nats.Msg)) error {
+	servers := "nats://nats1.polygon.io:30401, nats://nats2.polygon.io:30402, nats://nats3.polygon.io:30403"
+
+	nc, _ := nats.Connect(
+		servers,
+		nats.Token(apiKey))
+
+	_, err := nc.Subscribe("AM.*", handler)
```
Comment (umitanuki) [inline contrib/polygon/api/api.go]: How does it handle the error? Any retry?

Comment (rocketbitz) [reply contrib/polygon/api/api.go]: It reconnects on its own, you can only specify settings for max number of reconnects etc, and I am just leaving default. The package expects it to be used like this.

Comment (rocketbitz) [reply contrib/polygon/api/api.go]: I should however close the connection, and re-subscribe at the top level after some wait time.

File: contrib/polygon/polygon.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+	APIKey string `json:"api_key"`
+	// polygon API base URL in case it is being proxied
+	// (defaults to https://api.polygon.io/)
+	BaseURL string `json:"base_url"`
+}
+
+// NewBgWorker returns a new instances of PolygonFetcher. See FetcherConfig
+// for more details about configuring PolygonFetcher.
+func NewBgWorker(conf map[string]interface{}) (bgworker.BgWorker, error) {
+	data, _ := json.Marshal(conf)
```
Comment (ackleymi) [inline contrib/polygon/polygon.go]: Could ignoring this err ever go wrong? I wasn't sure.

File: contrib/polygon/polygon.go [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+	config := FetcherConfig{}
+	json.Unmarshal(data, &config)
```
Comment (ackleymi) [inline contrib/polygon/polygon.go]: Here too.

File: contrib/polygon/polygon.go (hunk 3 above)
Comment (umitanuki) [reply contrib/polygon/polygon.go]: This part is just to cast the same content from map to the concrete struct.  According to godoc, the only error case from `json.Marshal()` is when channel and other unsupported types are in the object tree. I don't think that's the case here.

**Comment by blonc** [comment]
getting an endless 'nats re-subscribing in 10s' error, API key is proper as i have tested it separately . 
 
```
I0613 13:25:45.437868       1 log.go:14] Trying to load module from path: /go/bin/stream.so...
I0613 13:25:45.487846       1 log.go:14] Success loading module /go/bin/stream.so.
I0613 13:25:45.487927       1 plugins.go:42] InitializeBgWorkers
I0613 13:25:45.487935       1 plugins.go:45] bgWorkerSetting = &{polygon.so Polygon map[api_key:_l_VN*************40t base_url:https://api.polygon.io]}
I0613 13:25:45.487968       1 log.go:14] Trying to load module from path: /go/bin/polygon.so...
I0613 13:25:45.509754       1 log.go:14] Success loading module /go/bin/polygon.so.
I0613 13:25:45.509852       1 plugins.go:51] Start running BgWorker Polygon...
I0613 13:25:45.509879       1 log.go:14] Launching heartbeat service...
I0613 13:25:45.509889       1 log.go:14] Enabling Query Access...
I0613 13:25:45.509893       1 log.go:14] Launching tcp listener for all services...
I0613 13:25:45.668452       1 polygon.go:57] nats re-subscribing in 10s
I0613 13:25:55.814518       1 polygon.go:57] nats re-subscribing in 10s
I0613 13:26:05.960775       1 polygon.go:57] nats re-subscribing in 10s
```

**Comment by rocketbitz** [comment]
You are right @blonc. It started happening after I added the unnecessary retry code for nats subscription. Have since fixed. Will be pushing soon.

File: contrib/polygon/polygon.go (hunk 4 above)
Comment (rocketbitz) [reply contrib/polygon/polygon.go]: what @umitanuki said.

**Comment by blonc** [comment]
@rocketbitz i also highly recommend putting in a symbols parameter, other wise you get the entire universe of something like 15-20k symbols, which is a massive amount of data.

**Comment by blonc** [comment]
getting the following now. checked the actual bin files as well and looks like nothing is writing.

```
I0613 18:46:46.505934       1 plugins.go:42] InitializeBgWorkers
I0613 18:46:46.505960       1 plugins.go:45] bgWorkerSetting = &{polygon.so Polygon map[api_key:_l_VN*********************0t base_url:https://api.polygon.io]}
I0613 18:46:46.506226       1 log.go:14] Trying to load module from path: /go/bin/polygon.so...
I0613 18:46:46.526262       1 log.go:14] Success loading module /go/bin/polygon.so.
I0613 18:46:46.526763       1 plugins.go:51] Start running BgWorker Polygon...
I0613 18:46:46.527025       1 log.go:14] Launching heartbeat service...
I0613 18:46:46.527049       1 log.go:14] Enabling Query Access...
I0613 18:46:46.527056       1 log.go:14] Launching tcp listener for all services...
E0613 18:47:26.271288       1 polygon.go:86] csm write failed (data shape does not match on-disk data shape: [{Epoch INT64} {Open FLOAT32} {High FLOAT32} {Low FLOAT32} {Close FLOAT32} {Volume INT32}] != [{Epoch INT64} {Open FLOAT64} {High FLOAT64} {Low FLOAT64} {Close FLOAT64} {Volume FLOAT64}])
E0613 18:47:59.559608       1 polygon.go:188] csm write failed (data shape does not match on-disk data shape: [{Epoch INT64} {Open FLOAT32} {High FLOAT32} {Low FLOAT32} {Close FLOAT32} {Volume INT32}] != [{Epoch INT64} {Open FLOAT64} {High FLOAT64} {Low FLOAT64} {Close FLOAT64} {Volume FLOAT64}])
E0613 18:51:01.649205       1 polygon.go:86] csm write failed (data shape does not match on-disk data shape: [{Epoch INT64} {Open FLOAT32} {High FLOAT32} {Low FLOAT32} {Close FLOAT32} {Volume INT32}] != [{Epoch INT64} {Open FLOAT64} {High FLOAT64} {Low FLOAT64} {Close FLOAT64} {Volume FLOAT64}])
E0613 18:51:23.243717       1 polygon.go:188] csm write failed (data shape does not match on-disk data shape: [{Epoch INT64} {Open FLOAT32} {High FLOAT32} {Low FLOAT32} {Close FLOAT32} {Volume INT32}] != [{Epoch INT64} {Open FLOAT64} {High FLOAT64} {Low FLOAT64} {Close FLOAT64} {Volume FLOAT64}])
```

**Comment by blonc** [comment]
i deleted `mktsdb` seems to have stopped the error, presume was just changes from one code version to another.

**Comment by rocketbitz** [comment]
@blonc I heard your issue w/ subscribing to everything. I didn't find anything in the polygon documentation for specifying a list of symbols to the NATS subscription, and I wasn't able to get it to work myself. I did however implement a basic filter with an input symbol list in the last commit. I think that should help a great deal with the load your system was experiencing.

**Comment by rocketbitz** [comment]
One important thing to note about this - if you are subscribing to all bars provided by polygon, and aggregating them using `aggtrigger.so`, the CPU load right now can get quite high and take several seconds to simmer down. This is due to the fact that all of the bars are written at almost exactly the same time, and so the aggtrigger runs for all of those symbols asynchronously, and queries the bars for the timeframes it needs to aggregate before flushing to disk. In my next PR, I will address this and optimize the aggtrigger to avoid this overhead.

**Comment by blonc** [comment]
@rocketbitz i am not to familiar with golang but something like the following works fine in python.

```
import asyncio
import os
import signal
import sys
import ast
import datetime
import json
from nats.aio.client import Client as NATS

# Be sure to: pip install asyncio-nats-client

def run(loop):
    nc = NATS()

    @asyncio.coroutine
    def closed_cb():
        print("Connection to NATS is closed.")
        yield from asyncio.sleep(0.1, loop=loop)
        loop.stop()
#nats://crypto1.polygon.io:30501
    options = {
        "servers": [
            "nats://{0}@nats1.polygon.io:30401".format('API_KEY_here'),
            "nats://{0}@nats2.polygon.io:30402".format('API_KEY_here'),
            "nats://{0}@nats3.polygon.io:30403".format('API_KEY_here')
        ],
        "io_loop": loop,
        "closed_cb": closed_cb
    }
    print( str(options) )

    yield from nc.connect(**options)
    print("Connected to NATS at {}...".format(nc.connected_url.netloc))

    @asyncio.coroutine
    def subscribe_handler(msg):
        subject = msg.subject
        reply = msg.reply
        data = json.loads(msg.data.decode())
        print(data,',')

#sym list
    sym = ["C.SEK_DKK","C.AUD_HKD"]
#loop
    for x in sym:
        yield from nc.subscribe(x, cb=subscribe_handler)


    def signal_handler():
        if nc.is_closed:
            return
        print("Disconnecting...")
        loop.create_task(nc.close())

    for sig in ('SIGINT', 'SIGTERM'):
        loop.add_signal_handler(getattr(signal, sig), signal_handler)

if __name__ == '__main__':
    loop = asyncio.get_event_loop()
    loop.run_until_complete(run(loop))
    try:
        loop.run_forever()
    finally:
        loop.close()
```

**Comment by rocketbitz** [comment]
Ah I see so a separate subscription is required for each. Makes sense. I will modify the plugin.

**Comment by blonc** [comment]
i am not sure if you can take the func stream() and do a loop over this bit

`_, err := nc.Subscribe("AM.*", handler)`

**Comment by rocketbitz** [comment]
Yes, that's the plan.

**Pull request merged**